### PR TITLE
Node 16

### PR DESCRIPTION
**Pull request by martindale**
Final updates to migrate to Node 16 as primary deploy target.

**TODO**
- [x] ~~Nix builds working (@anandsuresh / #79)~~
- [x] Nix builds working (@anandsuresh / #105 )
- [x] #93 (@martindale)

### CODE REVIEW

File: types/node.js [hunk 1]
```diff
@@ -51,22 +50,19 @@ class Node extends Service {
...
+      self.emit('debug', `[FABRIC:DEBUG] ${extra}${debug}`);
     });
 
     source.on('info', function (info) {
       console.log(`[FABRIC:INFO] ${extra}`, info);
     });
 
     source.on('log', function (log) {
-      console.log(`[FABRIC:LOG] ${extra}`, log);
+      self.emit('log', `[FABRIC:LOG] ${extra}${log}`);
```
Comment (anandsuresh) [inline types/node.js:65]: Is there a difference between `info` and `log` levels? IMO they are the same, but named differently, and in a monitoring scenario, they would be routed identically. Perhaps we pick one and go with that? I'd vote `info`, with `log` being set deprecated.

Comment (anandsuresh) [reply types/node.js:65]: Alternately, we can have a `log` event with a handler signature like `function (logLevel, ...args) { ... }`.  These events can then be wired into the logger to produce log out in the desired format.

One could even say that the `logger` class would contain a handler that can be wired up to a class instance as follows `instance.on('log', logger.onLogEvent)`.

File: scripts/validator.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env node
+'use strict';
+
+// Settings
+const settings = require('../settings/local');
+const master = Buffer.from('7e1b8e405919c1c8dcb07a677a2881b62517ecb7d136681851f514f22a806685', 'hex')
```
Comment (anandsuresh) [inline scripts/validator.js:6]: Curious what this constant is.

File: services/bitcoin.js [hunk 3]
```diff
@@ -18,11 +18,11 @@ const bip68 = require('bip68');
 const ECPair = ECPairFactory(ecc);
 
 // TODO: remove bcoin
-const bcoin = require('bcoin');
+// const bcoin = require('bcoin');
```
Comment (anandsuresh) [inline services/bitcoin.js:21]: I noticed a bunch of stuff commented out. Will these be removed before merge?

File: services/bitcoin.js [hunk 4]
```diff
@@ -33,20 +33,19 @@ const Service = require('../types/service');
...
-const KeyRing = bcoin.KeyRing;
-const MTX = bcoin.MTX;
-const NetAddress = bcoin.net.NetAddress;
-const Script = bcoin.Script;
+// const Amount = bcoin.Amount;
+// const Coin = bcoin.Coin;
+// const FullNode = bcoin.FullNode;
+// const MTX = bcoin.MTX;
+// const NetAddress = bcoin.net.NetAddress;
+// const Script = bcoin.Script;
```
Comment (anandsuresh) [inline services/bitcoin.js:48]: dead code removal?

**Review by anandsuresh — COMMENTED** [review]
First pass took a while, but I think this would be a good starting point. Happy to go over any and all comments on a call.

File: services/bitcoin.js (hunk 4 above)
Comment (martindale) [reply services/bitcoin.js:48]: Addressed in 11d4429d6a73eb4c50bea7d01999fc2bd3e4d8b1 (included with #93)

File: package.json [hunk 5]
```diff
@@ -95,55 +100,50 @@
     "glob-run": "^0.1.6",
     "javascript-state-machine": "^3.1.0",
     "jayson": "^3.4.4",
-    "js-beautify": "^1.10.0",
+    "js-beautify": "^1.14.3",
```
[the remark this answers is not in the document]
Comment (martindale) [reply package.json:103]: Nice catch, fixed in 944864040588f4211422cd1b7e06a1c2cf07b081

File: services/bitcoin.js (hunk 3 above)
Comment (martindale) [reply services/bitcoin.js]: Should be fixed now!

File: scripts/validator.js (hunk 2 above)
Comment (martindale) [reply scripts/validator.js]: hex-encoded master private key for the playnet identity (public)

**Comment by casey-bowman** [comment]
I was able to run fabric from this branch - for the 1st time! - from a pull a half-hour ago

I ran fabric setup and saw Wallet, Wallet State, and Wallet Keys. I then ran fabric start and see a repeated FABRIC:COMMIT with state, type, and id.

Earlier in the day I had found a way around an issue I had with a native build for the zeromq module on pop os on a librem 13 
(I ran "npm run build:libzmq" at node-modules/zeromq)

**Comment by martindale** [comment]
> I was able to run fabric from this branch - for the 1st time! - from a pull a half-hour ago
> 
> I ran fabric setup and saw Wallet, Wallet State, and Wallet Keys. I then ran fabric start and see a repeated FABRIC:COMMIT with state, type, and id.
> 
> Earlier in the day I had found a way around an issue I had with a native build for the zeromq module on pop os on a librem 13 (I ran "npm run build:libzmq" at node-modules/zeromq)

Thanks Casey!  Interested to hear @anandsuresh's opinion re: ZMQ inclusion here.  Main use is in `services/bitcoin.js` for live updates from a trusted Bitcoin node.

**Comment by anandsuresh** [comment]
@casey-bowman ZeroMQ tries to be smart about compilng and installing a native module, and that smartness tends to get in the way sometimes. I have found that starting from a clean slate (fresh clone of the fabric repo), I can get to a working state by running `nix-shell -p nodejs-16_x python3 zeromq --run "npm_config_zmq_external=true npm install"`. Curious to know if this solves your issue.

**Comment by casey-bowman** [comment]
> @casey-bowman ZeroMQ tries to be smart about compilng and installing a native module, and that smartness tends to get in the way sometimes. I have found that starting from a clean slate (fresh clone of the fabric repo, and checking out the `feature/node-16` branch), I can get to a working state by running `nix-shell -p nodejs-16_x python3 zeromq --run "npm_config_zmq_external=true npm install"`. Curious to know if this solves your issue.

Looking through my notes from yesterday, I see I had tried something like that - `npm install zeromq!5.2.8 --zmq-external`, which didn't work for me. I can try to use `npm_config_zmq_external="true"` next time and see, or even better yet the nix-shell command. Thanks, Anand

File: TODO.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
+# Things to Do
+- [ ] Test with Node 14
```
[the remark this answers is not in the document]
Comment (martindale) [reply TODO.md:10]: @melnx ok with Node 16 as minimum?

**Comment by casey-bowman** [comment]
I looked through the commits of this PR to get a sense of the recent code history, stopping here and there to look at some of the changes

**Pull request merged**